src/cudadecoder/cuda-online-pipeline-dynamic-batcher.cc

```diff
@@ -44,10 +51,76 @@ CudaOnlinePipelineDynamicBatcher::~CudaOnlinePipelineDynamicBatcher() {
 
 void CudaOnlinePipelineDynamicBatcher::Push(
     CorrelationID corr_id, bool is_first_chunk, bool is_last_chunk,
-    const SubVector<BaseFloat> &wave_samples) {
-  std::lock_guard<std::mutex> lk(chunks_m_);
-  chunks_.push_back({corr_id, is_first_chunk, is_last_chunk, wave_samples});
-  n_chunks_not_done_.fetch_add(1);
+    const SubVector<BaseFloat> &wave_samples_subv) {
+  std::lock_guard<std::mutex> lk(next_batch_and_backlog_m_);
+
+  // Try to add the chunk is the batch right away
+  if (!TryAddChunkToNextBatchDeepCopy(corr_id, is_first_chunk, is_last_chunk,
+                                      wave_samples_subv)) {
+    // If we cannot add the chunk to the next batch,
+    // put it in the backlog
+    Vector<BaseFloat> wave_samples(wave_samples_subv);  // deep copy
+    backlog_.push_back(
```

Comment:
1. Am I understanding that the backlog essentially holds all the data pushed so far and not fitting the batch, with an essentially unbounded growth? In my current design, I own the waveforms anyway.
2. The deep copying here seems to increase the time spent under the lock. Memory locality here is not great already, as the total amount of data pumped causes L2 misses, and L1 is not even in the game, so memory accesses are slow (this is unavoidable, I think). And this lock is a contention point for all streaming channels, and at the same time must be obtained by the BatcherThreadLoop every 100 us to avoid starving the device. This effect seems to be counteracted by the filling from the backlog in the same thread, without the need to obtain the log again. Does this provide an overall win for the contention on this single lock?

I'm wondering because my use is different from realtime streaming in the online example: I'm streaming much faster than realtime.

Reply:
Yes, that would behave poorly if each stream is pushing audio faster than realtime. The assumption here is that most chunks should be placed directly into the batch, skipping the backlog. But that doesn't work if we remove the "realtime" part. Do you own the waveforms because you own the entire audio file anyway ?

Reply:
I don't think this should behave poorly because of the backlog. Both sleeping and acquiring a lock are expensive operations, and backfilling happens while the lock is held. I really don't know right now, I'm only starting on perfing the whole pipeline right now. 

I do not own the entire file, it's streamed, but in quite hefty buffers. Also, there are segment in and out points, so every stream is jerky and contains multiple utterances often separated by silence. It's complicated.

All in all, let me play with it first; I'll certainly try many scenarios, and will keep you in the loop. The new code is simpler than the previous version, much easier to reason about the parts which are prone to locking contention, backlog blow-up etc., so I'll give it a good dusting! :)